### DISCUSSION TITLE
Replace "Friendly Date" format in favor of Date/TimeStamp

### DISCUSSION THREAD

**Discussion by ghost** [community/community#16550] [Other Feature Feedback, Questions, & Ideas]
I love the History, but what is very useless in the gui is how the History of anything is presented. Currently they are using the "Friendly" date format, which is vague and ambiguous, ie: "yesterday, few minutes ago, 2hours ago".
![image](https://user-images.githubusercontent.com/62229442/168346239-8c3b0442-2f47-43a0-a110-7a8dbd9dffe8.png)

On any page, the date is always in this useless friendly format and it is simply frustrating and makes working more cumbersome.
![image](https://user-images.githubusercontent.com/62229442/168345344-4b54c59e-7643-4a34-a571-a1015617a7a8.png)

In order to see the time I made the commit/change, I have to try to find the "sweet spot" for where to hover over (the whole list item is not triggering the hover...)
Date/Time should be in Standard format and not this unfriendly "Friendly" format.

Developers like to be concise, and being able to see actual time stamps next to commits can be extremely helpful when debugging or need to go back in history.
Please either let us choose date format in a settings under our profiles or replace this "friendly" date format with Standard Date/TimeStamp!
We don't want to have to do date arithmetics when trying to match our repos/work with our files on our computers.
If we have multiple of copies of the work on our computers, matching with dates only possible with literally doing date arithmetic.
If we could just see a TimeStamp: 05/10/2022 22:44:17, all guess work would be eliminated.

Also: The user should be able to decide on what date formatting works for them.  For example, the 24-hour (as above) format might not work for some people, and they might prefer to see AM/PM style format, other cultures might have different format that they are used to, and we all should be able to choose a formatting style for the date and time stamps.  These are very important pieces of data and not seeing exact values is frustrating and cumbersome.

Instead of forcing us to hover over the exact right area to see the timestamp: why not make that default??? Please consider this change!

**Comment by ExcuseLme** [comment]
can't agree anymore, I just want to see a timestamp, but not an abbreviated month names

**Comment by RasmusBruun** [comment]
I'd love it if this suggestion was implemented! 🔥

**Comment by ghost** [comment]
Hi all! 

I created an interesting concept based on your idea! What if there was a calendar in the sidebar where the user could, like a time machine, go back and forth with each modification of his repository?

#### concept/idea
![image](https://user-images.githubusercontent.com/111701513/187821294-5b095253-dfed-4713-a3b9-67f129ff8d9b.png)

**Reply by RobertKrawitz** [reply]
Calendars, hovering, and so forth are not useful if I want to cut and paste text or take a screenshot.

**Reply by GfEW** [reply]
I actually do like the calender / time machine idea, but it's entirely different from (and would likely take much more effort than) possible solutions to the simple datetime display format issue raised here.

**Comment by Calinou** [comment]
I don't think timestamps are an upgrade in most situations, unless you're taking a screenshot for posterity. As for hovering, I suppose a dummy element could be added around the date to make the hover area larger (and therefore easier to trigger).

What I'd like to see the most is some highlighting for recently changed files (e.g. by displaying the date in progressive shades of orange as the commit date approaches `now`). This could also apply to issues and comments – I find this beneficial in almost any situation where a relative date is displayed.

**Comment by lsehoon** [comment]
please please remove friendly dates please :pray:

**Reply by chaojian-zhang** [reply]
"Friendly dates" are such a hassle. BuildBot has it and it's equally annoying. Microsoft OneDrive also has it.

**Comment by lukas-wg2** [comment]
I have the same issue with the timestamp, but I don't necessarily think it should be removed. It should just be an option in your GitHub profile that you set like `Show relative time` and `Show absolute time`. For me, the biggest headache is the AM/PM as I'm used to 24-hour time format.

**Reply by timfrith** [reply]
I agree - I would prefer to have the choice in my settings of how to show dates.

**Reply by Mariella-Pasamonte** [reply]
Where can I find that option?

**Comment by diligent176** [comment]
Absolute date is what I want to know, 99% of the time.
Could we please have a checkbox for this in Profile / Appearance (near "default emoji skin tone")?
Date formats are far more important than emoji colors.  ;-p

@gabor-akeero

**Comment by maddes-b** [comment]
+1 for making this an option.

"1h ago" is not precise (can be anything from 60 to 119 minutes ago), and I don't like "1 months ago" or older.
The web page should just give coders/admins all detail information so they can work without hassle.
Using UTC as default timezone would be fine for coders/admins, and newbies can easily learn what's the delta to UTC via Wikipedia or [www.timeanddate.com](https://www.timeanddate.com/).

**Workaround:**
I installed [Tampermonkey](https://www.tampermonkey.net/) (a Greasemonkey alternative) on Firefox for UserScripts and then the UserScript [GitHub-static-time](https://github.com/Mottie/GitHub-userscripts/wiki/GitHub-static-time). That works well.
I am using the following format `Y-MM-DD HH:mm:ss z` (see [moment.js formats)](https://momentjs.com/docs/#/displaying/format/), [no UTC yet.](https://github.com/Mottie/GitHub-userscripts/issues/177) Thanks to [refined-github](https://github.com/refined-github/refined-github/issues/6143) for hinting me to this.

**Still the option to show absolute dates should be a GitHub standard feature.**

**Comment by RasmusBruun** [comment]
I've been working on a Chrome extension for this. 

In the current version you can toggle betweeen showing precise, "friendly" or both date formats. You can toggle between Am/Pm, and can show the precise time as either Short, Medium, Long format.

If there is interest in this, I'll go through the hassle of publising it :) (Apparently you have to pay a fee to make a Google Web Store Dev account 🙄) 
In the meantime you can find the repo and install it locally here: https://github.com/RasmusBruun/unfriendly-dates

**Reply by maddes-b** [reply]
@RasmusBruun Please read my comment above yours.

**Comment by mzso** [comment]
Date should be ISO time though. not that dumb (and ambiguous) format shown in the OP.

**Reply by aswen** [reply]
> Date should be ISO time though. not that dumb (and ambiguous) format shown in the OP.

I would prefer that too, but it's a matter of taste. It's not very friendly to the OP to describe your preference this way.

Wouldn't it be the best if it were **a configurable option**? One should be able to define it to their own liking. For example using the syntax that the [`date` command](https://www.gnu.org/software/coreutils/manual/html_node/Date-format-specifiers.html) supports.

So, @mzso and I would choose `%F %T`, @kahopkin would probably set it to either `%c` or `%B %d, %Y, %I:%M %p %Z` or so.
And if you don't set it: default to relative date/time.

**Reply by maddes-b** [reply]
That this should be options was mentioned several times in the previous comments: a) date format (ISO, etc.) and b) time zone (UTC)

**Reply by GfEW** [reply]
On absolute times, and ISO format specifically, there's also plenty discussion and workarounds in #5972.

**Comment by YCJDWCJI** [comment]


**Reply by mstockley-bladebridge** [reply]
This is pretty good, but 'az' doesn't seem to provide ISO format.  Also, it barely helps on Blame of large files that lazy-load lines as you scroll.

**Comment by thomas-maeder** [comment]
@MiaoMiao-Xie : thanks a lot!
Unfortunately, I see the "unfriendly" time format only for a fraction of a second, I think (it's hard to tell). After that, the (not so) "friendly" time format appears again.

**Reply by mstockley-bladebridge** [reply]
@thomas-maeder I'm actively working on a mutation-based version.  Give me a few minutes to finish testing and post it.  :)

**Reply by mstockley-bladebridge** [reply]
```javascript
// ==UserScript==
// @name        GitHub Absolute Dates
// @version     2023.12.20
// @match       *://github.com/*
// @match       *://*.github.com/*
// @icon        https://github.com/favicon.ico
// @description inspired by https://github.com/orgs/community/discussions/5972#discussioncomment-7766075 and https://stackoverflow.com/a/3552493/4808213
// ==/UserScript==
(function() {
	// constants
	var tz = Intl.DateTimeFormat().resolvedOptions().timeZone;
	var loc = navigator.language;
	var dd = '2-digit'; // shorthand

	// settings
	var format_hour12 = false;
	var format_parts = [
		// 'sep' is special, but any other documented parts are valid:
		// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat/DateTimeFormat#date-time_component_options
		{year: 'numeric'},
		{sep: '-'},
		{month: dd},
		{sep: '-'},
		{day: dd},
		{sep: ' '+(window.location.href.includes('/blame/') ? '&nbsp;' : '')},
		{hour: dd},
		{sep: ':'},
		{minute: dd},
		{sep: ':'},
		{second: dd},
		//{sep: ' '},
		//{timeZone: tz},
	];

	var parent = document.querySelector('#js-repo-pjax-container');
	if (!parent) return;

	var observer = new MutationObserver(function(mutations) {
		if (!mutations || !mutations.length) return;

		for (var m = 0; m < mutations.length; m++) {
			var mutation = mutations[m];
			var target = mutation.target;
			if (String(target.nodeName || '').toLowerCase() == 'relative-time') {
				replaceTime(target);
				continue;
			}

			var nodes = mutation.addedNodes;
			if (!nodes || !nodes.length) continue;

			for (var n = 0; n < nodes.length; n++) {
				target = nodes[n];
				if (String(target.nodeName || '').toLowerCase() == 'div') {
					replaceAllTimes(target);
				}
			}
		}
	});

	window.addEventListener('load', function(event) {
		replaceAllTimes(parent);

		observer.observe(parent, {
			attributes : true,
			childList : true,
			subtree : true,
		});
	});

	function formatDate(dateTime) {
		var date = new Date(dateTime);
		function format(option) {
			if ('sep' in option) {
				return option.sep;
			} else {
				option.timeZone = tz;
				option.hour12 = format_hour12;
				return date.toLocaleString(loc, option);
			}
		}
		return format_parts.map(format).join('');
	}

	function replaceAllTimes(parent) {
		if (!parent) parent = document;

		var elts = parent.querySelectorAll('relative-time');
		if (!elts || !elts.length) return;

		for (var e = 0; e < elts.length; e++) {
			replaceTime(elts[e]);
		}
	}

	function replaceTime(elt) {
		var dateTime = elt.getAttribute('datetime');
		var eltPadding = window.getComputedStyle(elt).getPropertyValue('padding-left');
		var eltTitle = elt.getAttribute('title');
		var parRelTime = elt.parentNode;
		if (parRelTime) {
			var span = document.createElement('span');
			span.setAttribute('datetime', dateTime);
			if (eltTitle) span.setAttribute('title', eltTitle);
			span.className = elt.className;
			span.innerHTML = formatDate(dateTime);
			parRelTime.replaceChild(span, elt);
			parRelTime.style.width = 'auto';
			span.style.whiteSpace = 'nowrap';
			if (!eltTitle) span.style.fontSize = 'smaller';
			if (eltPadding && (parseInt(eltPadding) != 0)) { // history
				span.style.paddingLeft = eltPadding;
			} else { // blame
				parRelTime.style.paddingRight = `${parseInt(window.getComputedStyle(parRelTime).getPropertyValue('padding-left')) || 10}px`;
			}
		}
	}
})();
```

**Reply by thomas-maeder** [reply]
Thanks! Almost there!
I get timestamps like this now: [2023-12-20 15 Uhr:01:54]
Without " Uhr" ("o'clock"), they would be correct.
That's with German (Swiss) language and 24h time format setting.

**Reply by thomas-maeder** [reply]
Oh, and Tampermonkey tells me that the meta-data of the new script are invalid. I had to create a mix of the old and new meta-data to test.

**Reply by mstockley-bladebridge** [reply]
My apologies for the bad meta-data; I'm testing in Violentmonkey and I admit I did not look up the requirements for Tampermonkey.  If you provide the valid headers, I'll be happy to update my post!

As for the " Uhr", I would not expect any locale to corrupt `2-digit` with a suffix, but locale-specific formatting is wild and crazy already so why not?  You can certainly change the items in `format_parts` to suit your needs; perhaps `{hour: 'numeric'}` will work as expected.

**Reply by thomas-maeder** [reply]
Something along the lines of the following works well for me:

	function formatDate(dateTime) {
		var date = new Date(dateTime);
        return new Intl.DateTimeFormat(navigator.languages,{
            year: "numeric",
            month: "numeric",
            day: "numeric",
            hour: "numeric",
            minute: "numeric",
            second: "numeric",
        }).format(date);
    }

**Reply by mstockley-bladebridge** [reply]
If only the parts were ordered sanely in my locale!  The only way I could find to specify the date order I want (year, month, day) is to format each piece separately and join them after (credit: [https://stackoverflow.com/questions/3552461/how-do-i-format-a-date-in-javascript/3552493#3552493](https://stackoverflow.com/questions/3552461/how-do-i-format-a-date-in-javascript/3552493#3552493)).  Once that gave me the array-map-join idea, I realized I could adapt it to inject any separator(s) anywhere I wanted; and so was born my overly-complex pair of `format_parts` array and `formatDate` function.  :)  Originally I had all the date pieces on one line and the time pieces on the next so I could easily toggle between date-time and just date while testing, but now I have moved each piece to its own line for the sake of readability.

@thomas-maeder Please do let me know what Tampermonkey disliked about my original set of meta-data.  I'd like to make sure the whole script is drop-in ready, but the documentation was not forthcoming about the requirements and restrictions!

**Reply by maddes-b** [reply]
Try the Tampermonkey solution mentioned in https://github.com/orgs/community/discussions/16550#discussioncomment-4160737

**Reply by thomas-maeder** [reply]
I have tried reproducing Tampermonkey's problem with the meta-data, but I haven't succeeded.
I must have messed up with the difficult task of copy&paste to cause it...

**Reply by mstockley-bladebridge** [reply]
@maddes-b I did, and it did not work on Blame pages because they lazy-load after an initial set of lines.  Believe me, I don't resort to building my own until I have to!  😆 

@thomas-maeder No worries, I think it may have been something I messed up when I originally posted without syntax highlighting.  If copy/paste works now, that's all that matters. :smiley:

**Reply by mzso** [reply]
It doesn't seem to work anymore. Do you happen to have an update?

**Comment by BrendanPC** [comment]
I want to add that not only would an absolute timestamp be preferable to a "friendly" date, currently the friendly date is a liar!

![image](https://github.com/community/community/assets/7386900/de8cf7d3-ce4d-41be-9cc2-3b7277f1dcd4)
This commit was made less than one year ago, but the friendly date claims it is two years old (i.e. it is off by over a year).

**Comment by aiqc** [comment]
it would be nice if you could control the date format using controls located near the column header. also sort by recent/old.

I'm sick of seeing "last month."  that's not helpful

**Reply by joann-alvarez** [reply]
Amen on sort by date option.

**Comment by radekroster** [comment]
"Friendly dates" makes me to search for the real date each time something happened more than 10 minutes ago. Nonsense.

**Comment by annbelievable** [comment]
Gitlab have a setting which lets user switches between a friendly date and actual date, the same could be achieved for github too.

**Comment by pkvdesigns** [comment]
Came here hoping for a solution, but instead made an Arc Boost to show the actual time below the Friendlytime:
https://arc.net/boost/604D04AC-915C-4C83-BEC8-023466BD96EC

Hope this helps some of you ;)

**Comment by GustavTaxen** [comment]
I cannot believe that it isn't possible to configure whether you want to see timestamps in UTC, local timezone, or "friendly" format in a service where dates/times are absolutely crucial for reporting and problem solving.

**Comment by anhofmann** [comment]
Today, I had to explain to the management the “friendly date” thing and that they need to hover with the mouse to see the real date. I can't describe the bewilderment in their faces, because to them, not seeing the exact dates reduces GitHub to an unorganized mess.

**Comment by codeaperature** [comment]
If you want a different format ... since the page element where the date occurs is: 
```
<td>
  <div class="react-directory-commit-age">
    <relative-time class="sc-aXZVg" tense="past" datetime="2024-10-25T22:54:17.000Z" title="Oct 25, 2024, 3:54 PM PDT">Oct 25, 2024</relative-time>
  </div>
</td>
```
The date could easily be changed in the browser's dev console with a bit of javascript or with a Greasemonkey script. I'd go with the datetime tag but truncating the end part.

**Comment by FearlessHyena** [comment]
Since it's been almost 3 years, a response from GitHub on whether this will or will not be implemented would be much appreciated

**Comment by GaryReckard** [comment]
+1 from me here. I would love it if this were a personal setting, perhaps under "Settings" => "Appearance", you could add a "Timestamps" setting. Defaulted to the current, relative 'time ago' format, but allowing us to opt-in for an absolute formatting (e.g., YYYY-MM-DD HH:mm:ss).

**Comment by FearlessHyena** [comment]
It's been almost 3 years since this discussion started. Some kind of update from the GitHub team would be really appreciated at this point

**Comment by stephenmontgomery** [comment]
What makes this so frustrating, if you refresh a GH list/page, you can actually see the date being displayed before the front end formats it to "x hours ago"

**Comment by partlythomas** [comment]
Token bump - It is crazy that this is not an option to switch on/off!

**Comment by abportela** [comment]
Bumping, just to keep the tradition here.

**Comment by mirabilos** [comment]
+1 but make the timestamp from the hover or display 24 hours

**Comment by prof-braino** [comment]
Is there any way to switch between friendly date and numeric date?  Hovering on friendly date is excellently annoying.

**Comment by beadon** [comment]
agreed. going to bump this again.  I want to see real timestamps not "last month".  I need the exact date and time.

**Reply by mirabilos** [reply]
Indeed!

And relaxed ISO 8601 as format (i.e. `2026-08-03 16:54`) as a **minimum**.
